File: src/api/app/api/routes/purchase.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_current_user
from app.db.session import get_session
from app.schemas.purchase import (
    AddLinkPayload,
    PurchaseItemCreate,
    PurchaseItemLinkRead,
    PurchaseItemRead,
    PurchaseItemUpdate,
)
from app.services.purchase import PurchaseItemNotFoundError, PurchaseService
from sqlmodel import Session

router = APIRouter(prefix="/purchase-items", tags=["purchase"], dependencies=[Depends(get_current_user)])


def _svc(session: Session = Depends(get_session)) -> PurchaseService:
    return PurchaseService(session)


def _to_read(item, links) -> PurchaseItemRead:
    return PurchaseItemRead(
        id=item.id,
        name=item.name,
        notes=item.notes,
        status=item.status,
        created_at=item.created_at,
        bought_at=item.bought_at,
        links=[PurchaseItemLinkRead.model_validate(l) for l in links],
    )


@router.get("", response_model=list[PurchaseItemRead])
def list_items(
    status: str | None = Query(default=None),
    keyword: str | None = Query(default=None),
    svc: PurchaseService = Depends(_svc),
) -> list[PurchaseItemRead]:
    return [_to_read(item, links) for item, links in svc.list(status=status, keyword=keyword)]


@router.post("", response_model=PurchaseItemRead, status_code=status.HTTP_201_CREATED)
def create_item(payload: PurchaseItemCreate, svc: PurchaseService = Depends(_svc)) -> PurchaseItemRead:
    item, links = svc.create(payload)
    return _to_read(item, links)


@router.get("/{item_id}", response_model=PurchaseItemRead)
def get_item(item_id: UUID, svc: PurchaseService = Depends(_svc)) -> PurchaseItemRead:
    try:
        item, links = svc.get(item_id)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return _to_read(item, links)


@router.patch("/{item_id}", response_model=PurchaseItemRead)
def update_item(item_id: UUID, payload: PurchaseItemUpdate, svc: PurchaseService = Depends(_svc)) -> PurchaseItemRead:
    try:
        item = svc.update(item_id, payload)
        _, links = svc.get(item_id)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return _to_read(item, links)


@router.delete("/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_item(item_id: UUID, svc: PurchaseService = Depends(_svc)) -> None:
    try:
        svc.delete(item_id)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e


@router.post("/{item_id}/buy", response_model=PurchaseItemRead)
def mark_bought(item_id: UUID, svc: PurchaseService = Depends(_svc)) -> PurchaseItemRead:
    try:
        item = svc.mark_bought(item_id)
        _, links = svc.get(item_id)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return _to_read(item, links)


@router.post("/{item_id}/unbuy", response_model=PurchaseItemRead)
def mark_to_buy(item_id: UUID, svc: PurchaseService = Depends(_svc)) -> PurchaseItemRead:
    try:
        item = svc.mark_to_buy(item_id)
        _, links = svc.get(item_id)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return _to_read(item, links)


@router.post("/{item_id}/links", response_model=PurchaseItemLinkRead, status_code=status.HTTP_201_CREATED)
def add_link(item_id: UUID, payload: AddLinkPayload, svc: PurchaseService = Depends(_svc)) -> PurchaseItemLinkRead:
    try:
        link = svc.add_link(item_id, payload)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return PurchaseItemLinkRead.model_validate(link)


@router.delete("/{item_id}/links/{link_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_link(item_id: UUID, link_id: UUID, svc: PurchaseService = Depends(_svc)) -> None:
    try:
        svc.remove_link(link_id)
    except PurchaseItemNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e

File: src/web/src/api/purchase.ts
import { supabase } from "../lib/supabase";

const BASE = import.meta.env.VITE_API_URL || "/api/v1";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const { data: { session } } = await supabase.auth.getSession();
  const token = session?.access_token;
  const res = await fetch(`${BASE}${path}`, {
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...init?.headers,
    },
    ...init,
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw Object.assign(new Error(body.detail ?? res.statusText), { status: res.status });
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

export interface PurchaseLink {
  id: string;
  url: string;
  created_at: string;
}

export interface PurchaseItem {
  id: string;
  name: string;
  notes: string | null;
  status: "to_buy" | "bought";
  created_at: string;
  bought_at: string | null;
  links: PurchaseLink[];
}

export interface CreatePurchasePayload {
  name: string;
  notes?: string | null;
  links?: string[];
}

export interface UpdatePurchasePayload {
  name?: string;
  notes?: string | null;
}

export const purchaseApi = {
  list: (status?: "to_buy" | "bought", keyword?: string) => {
    const params = new URLSearchParams();
    if (status) params.set("status", status);
    if (keyword) params.set("keyword", keyword);
    const qs = params.toString();
    return request<PurchaseItem[]>(`/purchase-items${qs ? `?${qs}` : ""}`);
  },

  get: (id: string) => request<PurchaseItem>(`/purchase-items/${id}`),

  create: (payload: CreatePurchasePayload) =>
    request<PurchaseItem>("/purchase-items", { method: "POST", body: JSON.stringify(payload) }),

  update: (id: string, payload: UpdatePurchasePayload) =>
    request<PurchaseItem>(`/purchase-items/${id}`, { method: "PATCH", body: JSON.stringify(payload) }),

  delete: (id: string) =>
    request<void>(`/purchase-items/${id}`, { method: "DELETE" }),

  markBought: (id: string) =>
    request<PurchaseItem>(`/purchase-items/${id}/buy`, { method: "POST" }),

  markToBuy: (id: string) =>
    request<PurchaseItem>(`/purchase-items/${id}/unbuy`, { method: "POST" }),

  addLink: (id: string, url: string) =>
    request<PurchaseLink>(`/purchase-items/${id}/links`, {
      method: "POST",
      body: JSON.stringify({ url }),
    }),

  removeLink: (id: string, linkId: string) =>
    request<void>(`/purchase-items/${id}/links/${linkId}`, { method: "DELETE" }),
};
